src/protocol/attribution/mod.rs

```diff
@@ -36,49 +38,113 @@ where
             .await?)
 }
 
-async fn compute_stop_bit<F, C, S>(
+///
+/// Computes `SUM(credits[i] through credits[i + n])` where `n` is the number of "matching rows", as indicated by the `helper_bits`
+/// This result is saved as credits[i].
+///
+/// Helper bits should be a sharing of either `1` or `0` for each row, indicating if that row "matches" the row preceeding it.
+///
+/// ## Errors
+/// Fails if the multiplication protocol fails.
+///
+pub async fn do_the_binary_tree_thing<F, C, S>(
     ctx: C,
-    record_id: RecordId,
-    b_bit: &S,
-    sibling_stop_bit: &S,
-    first_iteration: bool,
-) -> Result<S, Error>
+    helper_bits: &[S],
+    credits: &mut [S],
+) -> Result<(), Error>
 where
     F: Field,
     C: Context<F, Share = S>,
-    S: SecretSharing<F>,
+    S: ArithmeticSecretSharing<F>,
 {
-    // This method computes `b == 1 ? sibling_stop_bit : 0`.
-    // Since `sibling_stop_bit` is initialize with 1, we return `b` if this is
-    // the first iteration.
-    if first_iteration {
-        return Ok(b_bit.clone());
+    let num_rows = helper_bits.len() + 1;
+
+    let depth_0_ctx = ctx
+        .narrow(&InteractionPatternStep::from(0))
+        .set_total_records(num_rows - 1);
+    let credit_updates = try_join_all(helper_bits.iter().enumerate().map(|(i, helper_bit)| {
+        let c = depth_0_ctx.clone();
+        let record_id = RecordId::from(i);
+        let credit = &credits[i + 1];
+        async move { c.multiply(record_id, helper_bit, credit).await }
+    }))
+    .await?;
+    credit_updates
+        .into_iter()
+        .enumerate()
+        .for_each(|(i, credit)| {
+            credits[i] += &credit;
+        });
```

Comment:
We can merge these two iterations into one like below. That makes the input not `&mut` and return a new `Vec` instead, but we are creating a new vector `credit_updates` anyway. So, removing another iteration is slightly better?

```
    let credit_updates = try_join_all(
        std::iter::zip(credits.iter(), credits[1..].iter())
            .zip(helper_bits)
            .enumerate()
            .map(|(i, ((current_credit, next_credit), helper_bit))| {
                let c = depth_0_ctx.clone();
                let record_id = RecordId::from(i);
                // let credit = &credits[i + 1];
                async move {
                    Ok::<_, Error>(
                        current_credit.clone()
                            + &c.multiply(record_id, helper_bit, next_credit).await?,
                    )
                }
            }),
    )
    .await?;
```